### ISSUE TITLE
Better literate programming

### ISSUE THREAD

**Issue by denismerigoux** [CatalaLang/catala#29]
## Context

Right now, the surface Catala AST mixes code items as well as literate programming headings and items : https://github.com/CatalaLang/catala/blob/master/src/catala/catala_surface/catala_ast.ml#L192. This makes it uneasy to determine to which "legislative atom" a chunk of code belongs to.

Indeed, if we are to compile Catala programs into programs that can explain why they got to their results, we should be able to link a position in a code block with a reference to a legislative atom: an article for French text, a code section for English texts, etc.

## Goal

In order to do that, we need to change the flat structure of `program_item` into a tree-shaped structure where the headings and atom title would be nodes of the tree, and the code blocks would be the leaves. Then, when desugaring the AST, we can augment `Pos.t` with the additional information like `Legislative Code XXX -- Section YYY --- Paragraph ZZZ -- ...`

## What to do ?

The main mission is to refactor `Catala_ast.program_item` into something more like : 

```OCaml
type program_item = 
  | CodeBlock of code_block 
  | LawTex of law_text

type program_node = 
  | LawHeading of string * int * program_item list
  | LawArticle of law_article * program_item list 
```

Then, the parser and especially this rule https://github.com/CatalaLang/catala/blob/master/src/catala/catala_surface/parser.mly#L500 should be updated accordingly. Finally, `Pos.t` should be augmented with the information of the path of `program_nodes` leading to the chunk of code.

**Comment by denismerigoux** [comment]
Fixed by cdfa9038cf78a8000984ce6943faa34121bd4d28. But now there is a shift-reduce conflict in Menhir due to law text appearing in positions : inside a law article block OR between a heading an an article title. Precedences should be given to solve the conflict in favor in shifting but I never could find out how to give precedence correctly in Menhir :)

**Comment by denismerigoux** [comment]
Menhir warnings removed by 40a4f90. What's left to do is modify `Pos.t` to include the path inside the legislative hierarchical structure where the snippet of code is.
